### PR TITLE
Add support for multiple normalizers

### PR DESCRIPTION
**Pull request by kou**
GitHub: fix GH-955

This should be backward compatible. It means that newer Groonga can
use DB created by old Groonga but old Groonga can't use DB created by
new Groonga properly.

Old:

  * Normalizer ID is stored in table's header.

New:

  * Normalizer IDs are stored in specs. (${DB}.0000000)

normalize command and table_create command accepts "--normalizers"
option. "--normalizer" option is also accepted for backward
compatibility. Users can specify multiple normalizers by
"--normalizer".

dump command, table_list command and schema command supports multiple
normalizers. dump command and schema command still use "--normalizer"
option not "--normalizers" option to support restoring dump created by
new Groonga with old Groonga.

New APIs:

  * grn_info_type::GRN_INFO_NORMALIZERS
  * grn_string_get_normalizer_index()
  * grn_table_set_normalizers_options()
  * grn_table_get_normalizers_options()
  * grn_table_cache_normalizers_options()
  * grn_table_get_normalizers_string()
  * grn_table_set_token_filters_options()
    This deprecates grn_table_set_token_filter_options().
  * grn_table_get_token_filter_options()
    This deprecates grn_table_get_token_filter_options().
  * grn_table_get_token_filters_options()
    This deprecates grn_table_get_token_filter_options().
  * grn_table_cache_token_filters_options()
    This deprecates grn_table_cache_token_filter_options().

### CODE REVIEW

**Comment by kou** [comment]
Cc: @HashidaTKS

**Review by HashidaTKS — COMMENTED** [review]
Results may vary depending on the order of normalizers  because all the normalizers are applied sequential.
Is my understanding correct ?

This is just a question about the specification, not requesting any changes.

File: lib/normalizer.c [hunk 1]
```diff
@@ -2373,12 +2374,13 @@ nfkc121_next(grn_ctx *ctx, int nargs, grn_obj **args, grn_user_data *user_data)
 
   table = grn_string_get_table(ctx, string);
   if (table) {
-    options = grn_table_cache_normalizer_options(ctx,
-                                                 table,
-                                                 string,
-                                                 nfkc121_open_options,
-                                                 nfkc121_close_options,
-                                                 NULL);
+    int32_t i = grn_string_get_normalizer_index(ctx, string);
```
[the remark this answers is not in the document]
Comment (kou) [reply lib/normalizer.c]: Good catch!
Applied.

File: lib/string.c [hunk 2]
```diff
@@ -243,33 +244,122 @@ grn_string_open_(grn_ctx *ctx,
   }
 
   {
-    grn_obj *normalizer = NULL;
+    grn_obj normalizers;
+    GRN_PTR_INIT(&normalizers, GRN_OBJ_VECTOR, GRN_ID_NIL);
     if (lexicon_or_normalizer) {
```
[the remark this answers is not in the document]
Comment (kou) [reply lib/string.c]: Good catch!
Removed.

File: lib/string.c [hunk 3]
```diff
@@ -243,33 +244,122 @@ grn_string_open_(grn_ctx *ctx,
   }
 
   {
```
[the remark this answers is not in the document]
Comment (kou) [reply lib/string.c:246]: Yes. There are 3 braces usages in Groonga:

  1. Define variables in the middle of function
  2. Represent a semantic block
  3. Restrict variable scope

Historically, we didn't require C99 before.
Codes written in the period use braces for 1., 2. and 3.

Codes written after the period use braces for 2. and 3.

For example, https://github.com/groonga/groonga/pull/1220/files#diff-51b979b3a1f20747d06a13d327390c8c1854e58c8dd5552595e8bf6c039606b6R642-R659 uses braces for 2. They represents "map" output codes.

**Review by kou — COMMENTED** [review]
> Results may vary depending on the order of normalizers because all the normalizers are applied sequential.
> Is my understanding correct ?

Correct.

File: lib/string.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (HashidaTKS) [reply lib/string.c:246]: I see. thank you!

**Comment by HashidaTKS** [comment]
LGTM

**Review by HashidaTKS — APPROVED** [review]

**Pull request merged**